crates/bevy_math/src/curve/interval.rs

```diff
@@ -0,0 +1,330 @@
+//! The [`Interval`] type for nonempty intervals used by the [`Curve`](super::Curve) trait.
+
+use itertools::Either;
+use std::{
+    cmp::{max_by, min_by},
+    ops::RangeInclusive,
+};
+use thiserror::Error;
+
+#[cfg(feature = "bevy_reflect")]
+use bevy_reflect::Reflect;
+#[cfg(all(feature = "serialize", feature = "bevy_reflect"))]
+use bevy_reflect::{ReflectDeserialize, ReflectSerialize};
+
+/// A nonempty closed interval, possibly infinite in either direction.
+#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(Reflect), reflect(Debug, PartialEq))]
+#[cfg_attr(
+    all(feature = "serialize", feature = "bevy_reflect"),
+    reflect(Serialize, Deserialize)
+)]
+pub struct Interval {
+    start: f32,
+    end: f32,
+}
+
+/// An error that indicates that an operation would have returned an invalid [`Interval`].
+#[derive(Debug, Error)]
+#[error("The resulting interval would be invalid (empty or with a NaN endpoint)")]
+pub struct InvalidIntervalError;
+
+/// An error indicating that an infinite interval was used where it was inappropriate.
+#[derive(Debug, Error)]
+#[error("This operation does not make sense in the context of an infinite interval")]
+pub struct InfiniteIntervalError;
+
+impl Interval {
+    /// Create a new [`Interval`] with the specified `start` and `end`. The interval can be infinite
+    /// but cannot be empty and neither endpoint can be NaN; invalid parameters will result in an error.
+    #[inline]
+    pub fn new(start: f32, end: f32) -> Result<Self, InvalidIntervalError> {
+        if start >= end || start.is_nan() || end.is_nan() {
+            Err(InvalidIntervalError)
+        } else {
+            Ok(Self { start, end })
+        }
+    }
+
+    /// Get the start of this interval.
+    #[inline]
+    pub fn start(self) -> f32 {
+        self.start
+    }
+
+    /// Get the end of this interval.
+    #[inline]
+    pub fn end(self) -> f32 {
+        self.end
+    }
+
+    /// Create an [`Interval`] by intersecting this interval with another. Returns an error if the
+    /// intersection would be empty (hence an invalid interval).
+    pub fn intersect(self, other: Interval) -> Result<Interval, InvalidIntervalError> {
+        let lower = max_by(self.start, other.start, |x, y| x.partial_cmp(y).unwrap());
+        let upper = min_by(self.end, other.end, |x, y| x.partial_cmp(y).unwrap());
+        Self::new(lower, upper)
+    }
+
+    /// Get the length of this interval. Note that the result may be infinite (`f32::INFINITY`).
+    #[inline]
+    pub fn length(self) -> f32 {
+        self.end - self.start
+    }
+
+    /// Returns `true` if both endpoints of this interval are finite.
+    #[inline]
+    pub fn is_finite(self) -> bool {
```

Comment:
maybe better than "(in)finite" throughout these changes is "(un)bounded".

"unbounded start" makes more intuitive sense to me than "infinite left endpoint".

Reply:
The name of the function fits imo because it's about the length of the interval. is_finite returns true if the interval is of finite length. If it's (-inf; 0), then the length is infinite.

And the finiteness of endpoints corresponds to the f32 actually being finite/infinite. f32 being bounded is... also possible I guess?

Boundedness is an equivalent property.

Just bikeshedding here.

Reply:
imo, if this were the `IntervalLength` type, i'd be inclined to agree, but `Interval` doesn't (shouldn't?) be represented, even conceptually, as a length.

`has_finite_length`, then would be better than `is_finite`, but i think `is_bounded` is best.

The `Interval` type is a nice usability improvement over the `RangeFoo` types for this kind of task, but there's no reason we should shy away from [their nomenclature](https://doc.rust-lang.org/stable/std/ops/struct.Range.html).

Reply:
I would be okay with changing some of the nomenclature to use "bounded" consistently. The reason I chose "infinite" is that I felt like it was more layman-friendly despite being less precise (i.e. to me, "bounded" is more like the mathematical terminology).

Reply:
I suppose if we wanted full mathematical precision, we would have `is_compact`. I like `is_bounded` but if we are going to bike-shed it maybe @alice-i-cecile would like input?

Reply:
I'm not sure what it is ambiguous with, but I don't feel strongly against `is_bounded` (although `is_finite` has my slight personal preference).